File: src/ai/input_processor.js
/**
 * Input Processor - Complete information encoding system
 * 
 * New Architecture:
 * - 50 boid vectors + 1 predator vector = 51 total entities
 * - Each entity: 4 features = 204 total inputs
 * - No vision limitations - all boids included
 * - Device-independent normalization for consistency
 * 
 * Boid Vector Format: [rel_x, rel_y, vel_x, vel_y]
 * Predator Vector Format: [canvas_width_norm, canvas_height_norm, vel_x, vel_y]
 */
function InputProcessor() {
    // Use centralized constants for new encoding system
    this.maxBoids = window.SIMULATION_CONSTANTS.MAX_BOIDS;
    this.boidVectorSize = window.SIMULATION_CONSTANTS.BOID_VECTOR_SIZE;
    this.predatorVectorSize = window.SIMULATION_CONSTANTS.PREDATOR_VECTOR_SIZE;
    this.totalInputSize = window.SIMULATION_CONSTANTS.NEURAL_INPUT_SIZE;
    
    // Unified velocity normalization - use the maximum of boid and predator speeds
    // This ensures all velocities are normalized consistently
    this.unifiedMaxVelocity = Math.max(
        window.SIMULATION_CONSTANTS.BOID_MAX_SPEED,
        window.SIMULATION_CONSTANTS.PREDATOR_MAX_SPEED
    );
    
    // Device-independent normalization bounds from centralized config
    this.maxDistance = window.SIMULATION_CONSTANTS.MAX_DISTANCE;
}

/**
 * Convert complete game state to neural network inputs
 * @param {Array} boids - Array of all boid objects
 * @param {Object} predatorPos - Predator position {x, y}
 * @param {Object} predatorVel - Predator velocity {x, y}
 * @param {number} canvasWidth - Canvas width
 * @param {number} canvasHeight - Canvas height
 * @returns {Array} 204-element input vector (51 entities × 4 features)
 */
InputProcessor.prototype.processInputs = function(boids, predatorPos, predatorVel, canvasWidth, canvasHeight) {
    var inputs = new Array(this.totalInputSize);
    var inputIndex = 0;
    
    // Encode all boid vectors (50 slots)
    for (var i = 0; i < this.maxBoids; i++) {
        if (i < boids.length) {
            // Encode existing boid
            var boidVector = this.encodeBoid(boids[i], predatorPos, canvasWidth, canvasHeight);
            for (var j = 0; j < this.boidVectorSize; j++) {
                inputs[inputIndex++] = boidVector[j];
            }
        } else {
            // Pad with zeros for missing boids
            for (var j = 0; j < this.boidVectorSize; j++) {
                inputs[inputIndex++] = 0.0;
            }
        }
    }
    
    // Encode predator vector (1 slot)
    var predatorVector = this.encodePredator(predatorVel, canvasWidth, canvasHeight);
    for (var j = 0; j < this.predatorVectorSize; j++) {
        inputs[inputIndex++] = predatorVector[j];
    }
    
    return inputs;
};

/**
 * Encode a single boid as a 4-feature vector
 * @param {Object} boid - Boid object with position and velocity
 * @param {Object} predatorPos - Predator position for relative calculation
 * @param {number} canvasWidth - Canvas width for edge wrapping calculation
 * @param {number} canvasHeight - Canvas height for edge wrapping calculation
 * @returns {Array} [rel_x, rel_y, vel_x, vel_y] - positions normalized, velocities clamped
 */
InputProcessor.prototype.encodeBoid = function(boid, predatorPos, canvasWidth, canvasHeight) {
    // Calculate relative position with edge wrapping support
    var relativePos = this.calculateRelativePosition(
        boid.position, predatorPos, canvasWidth, canvasHeight
    );
    
    // Normalize relative position by fixed maximum distance (no clamping - keep real spatial info)
    var relX = relativePos.x / this.maxDistance;
    var relY = relativePos.y / this.maxDistance;
    
    // Normalize and clamp velocity by unified max velocity (velocities have natural bounds)
    var velX = this.clamp(boid.velocity.x / this.unifiedMaxVelocity);
    var velY = this.clamp(boid.velocity.y / this.unifiedMaxVelocity);
    
    return [relX, relY, velX, velY];
};

/**
 * Encode predator as a 4-feature vector with world context and velocity
 * @param {Object} predatorVel - Predator velocity {x, y}
 * @param {number} canvasWidth - Canvas width
 * @param {number} canvasHeight - Canvas height
 * @returns {Array} [canvas_width_norm, canvas_height_norm, vel_x, vel_y] - canvas size normalized, velocities clamped
 */
InputProcessor.prototype.encodePredator = function(predatorVel, canvasWidth, canvasHeight) {
    // Normalize canvas dimensions (no clamping - keep real world boundary info)
    // This gives the network context about world boundaries
    var canvasWidthNorm = canvasWidth / this.maxDistance;
    var canvasHeightNorm = canvasHeight / this.maxDistance;
    
    // Normalize and clamp velocity by unified max velocity (velocities have natural bounds)
    // Keep velocity in positions 2,3 like boids for consistency
    var velX = this.clamp(predatorVel.x / this.unifiedMaxVelocity);
    var velY = this.clamp(predatorVel.y / this.unifiedMaxVelocity);
    
    return [canvasWidthNorm, canvasHeightNorm, velX, velY];
};

/**
 * Calculate relative position with edge wrapping support
 * @param {Object} boidPos - Boid position {x, y}
 * @param {Object} predatorPos - Predator position {x, y}
 * @param {number} canvasWidth - Canvas width
 * @param {number} canvasHeight - Canvas height
 * @returns {Object} Relative position {x, y}
 */
InputProcessor.prototype.calculateRelativePosition = function(boidPos, predatorPos, canvasWidth, canvasHeight) {
    // Calculate basic relative position
    var dx = boidPos.x - predatorPos.x;
    var dy = boidPos.y - predatorPos.y;
    
    // Handle edge wrapping (toroidal topology)
    // Choose the shorter distance considering wrapping
    if (Math.abs(dx) > canvasWidth / 2) {
        dx = dx > 0 ? dx - canvasWidth : dx + canvasWidth;
    }
    if (Math.abs(dy) > canvasHeight / 2) {
        dy = dy > 0 ? dy - canvasHeight : dy + canvasHeight;
    }
    
    return { x: dx, y: dy };
};

/**
 * Clamp value to [-1, 1] range
 * @param {number} value - Value to clamp
 * @returns {number} Clamped value
 */
InputProcessor.prototype.clamp = function(value) {
    return Math.max(-1, Math.min(1, value));
};

 

File: simulation/processors/input_processor.py
"""
Input Processor - Transformer-ready format for neural network inputs

This is a data conversion layer that transforms game state into structured inputs
for the neural network. This MUST match exactly with the JavaScript implementation.
"""

import math
from typing import List, Dict, Any

from config.constants import CONSTANTS

class InputProcessor:
    def __init__(self):
        # Use centralized constants (matches JavaScript)
        self.max_distance = CONSTANTS.MAX_DISTANCE
        
        # Unified velocity normalization - use the maximum of boid and predator speeds
        self.unified_max_velocity = max(
            CONSTANTS.BOID_MAX_SPEED,
            CONSTANTS.PREDATOR_MAX_SPEED
        )
    
    def process_inputs(self, boids: List[Dict], predator_pos: Dict, predator_vel: Dict, 
                      canvas_width: float, canvas_height: float) -> Dict[str, Any]:
        """
        Convert game state to transformer-friendly format
        
        Args:
            boids: List of boid objects with position and velocity
            predator_pos: Predator position {'x': float, 'y': float}
            predator_vel: Predator velocity {'x': float, 'y': float}
            canvas_width: Canvas width
            canvas_height: Canvas height
            
        Returns:
            Structured input with context, predator, and boids arrays
        """
        # Context information for world boundary awareness
        context = {
            'canvasWidth': canvas_width / self.max_distance,  # Normalized canvas width
            'canvasHeight': canvas_height / self.max_distance  # Normalized canvas height
        }
        
        # Predator velocity information
        predator = {
            'velX': self.clamp(predator_vel['x'] / self.unified_max_velocity),
            'velY': self.clamp(predator_vel['y'] / self.unified_max_velocity)
        }
        
        # Dynamic array of boids with relative positions and velocities
        boid_array = []
        for boid in boids:
            boid_data = self.encode_boid(boid, predator_pos, canvas_width, canvas_height)
            boid_array.append({
                'relX': boid_data[0],
                'relY': boid_data[1],
                'velX': boid_data[2],
                'velY': boid_data[3]
            })
        
        return {
            'context': context,
            'predator': predator,
            'boids': boid_array
        }
    
    def encode_boid(self, boid: Dict, predator_pos: Dict, canvas_width: float, canvas_height: float) -> List[float]:
        """
        Encode a single boid as relative position and velocity
        
        Args:
            boid: Boid object with position and velocity
            predator_pos: Predator position for relative calculation
            canvas_width: Canvas width for edge wrapping calculation
            canvas_height: Canvas height for edge wrapping calculation
            
        Returns:
            [rel_x, rel_y, vel_x, vel_y] - normalized values
        """
        # Calculate relative position with edge wrapping support
        relative_pos = self.calculate_relative_position(
            boid['position'], predator_pos, canvas_width, canvas_height
        )
        
        # Normalize relative position by fixed maximum distance
        rel_x = relative_pos['x'] / self.max_distance
        rel_y = relative_pos['y'] / self.max_distance
        
        # Normalize and clamp velocity by unified max velocity
        vel_x = self.clamp(boid['velocity']['x'] / self.unified_max_velocity)
        vel_y = self.clamp(boid['velocity']['y'] / self.unified_max_velocity)
        
        return [rel_x, rel_y, vel_x, vel_y]
    
    def calculate_relative_position(self, boid_pos: Dict, predator_pos: Dict, 
                                  canvas_width: float, canvas_height: float) -> Dict[str, float]:
        """
        Calculate relative position with edge wrapping support
        
        Args:
            boid_pos: Boid position {'x': float, 'y': float}
            predator_pos: Predator position {'x': float, 'y': float}
            canvas_width: Canvas width
            canvas_height: Canvas height
            
        Returns:
            Relative position {'x': float, 'y': float}
        """
        # Calculate basic relative position
        dx = boid_pos['x'] - predator_pos['x']
        dy = boid_pos['y'] - predator_pos['y']
        
        # Handle edge wrapping (toroidal topology)
        # Choose the shorter distance considering wrapping
        if abs(dx) > canvas_width / 2:
            dx = dx - canvas_width if dx > 0 else dx + canvas_width
        if abs(dy) > canvas_height / 2:
            dy = dy - canvas_height if dy > 0 else dy + canvas_height
        
        return {'x': dx, 'y': dy}
    
    def clamp(self, value: float) -> float:
        """Clamp value to [-1, 1] range"""
        return max(-1, min(1, value)) 